### PR TITLE
fix: Add special cases for `clone` and `to_copy` where input of graph is output

### PR DESCRIPTION
**Pull request by gs-olive**
# Description

- TRT does not allow inputs of graphs to be outputs as well, however many of the scenarios encountered in real models can have this situation come up, especially in cases where the input is cloned or copied and then returned
- The current converters will register these operators as a no-op, causing TRT engine building to fail on such inputs
- Instead of requiring creation of an identity layer for every case of a clone or copy node, we instead check if that node is the only operator on a placeholder (input) and then insert the identity layer or not, accordingly
- Coalesce implementations of clone and to_copy, which are effectively the same operator
- Add test cases to validate new behavior
- Add new boilerplate converter validator utility to support this case

Addresses bug in #1565

## Type of change

- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)
# Checklist:

- [ x ] My code follows the style guidelines of this project (You can use the linters)
- [ x ] I have performed a self-review of my own code
- [ x ] I have commented my code, particularly in hard-to-understand areas and hacks
- [ x ] I have made corresponding changes to the documentation
- [ x ] I have added tests to verify my fix or my feature
- [ x ] New and existing unit tests pass locally with my changes
- [ x ] I have added the relevant labels to my PR in so that relevant reviewers are notified

### CODE REVIEW

File: py/torch_tensorrt/dynamo/conversion/aten_ops_converters.py [hunk 1]
```diff
@@ -401,28 +432,41 @@ def aten_ops_to_copy_dtype(
...
+    # we need to force cast to ensure a layer is added to the TRT engine
+    # since TRT engine inputs cannot also be TRT engine outputs
+    return impl.cast.to_copy(
         network,
         target,
         SourceIR.ATEN,
         name,
         args[0],
+        kwargs.get("dtype", args[0].dtype),
+        force_cast=True,
```
Comment (gs-olive) [inline py/torch_tensorrt/dynamo/conversion/aten_ops_converters.py:465]: Rename to `force_layer`

File: py/torch_tensorrt/dynamo/conversion/aten_ops_converters.py [hunk 2]
```diff
@@ -401,28 +432,41 @@ def aten_ops_to_copy_dtype(
...
     target: Target,
     args: Tuple[Argument, ...],
     kwargs: Dict[str, Argument],
     name: str,
 ) -> Union[TRTTensor, Sequence[TRTTensor]]:
-    return impl.cast.clone(
+    # For clone or copy nodes where the input is also the output,
+    # we need to force cast to ensure a layer is added to the TRT engine
+    # since TRT engine inputs cannot also be TRT engine outputs
+    return impl.cast.to_copy(
```
Comment (gs-olive) [inline py/torch_tensorrt/dynamo/conversion/aten_ops_converters.py:458]: Add logic to `impl.cast` to perform differentiation on identity layers

**Comment by narendasan** [comment]
@gs-olive is this only for graphs like: 
```
graph(x: Tensor):
  return x
```

Is it worth even converting these into engines?

**Comment by gs-olive** [comment]
@narendasan - This change is primarily for cases encountered in detectron-style models where inputs are also outputs of the engine. It might be something like:
```python
graph(x: Tensor, y: Tensor):
  # A lot of logic/operations applied to y
  ...
  return x, y_new
```

File: py/torch_tensorrt/dynamo/conversion/impl/elementwise/base.py [hunk 3]
```diff
@@ -116,9 +121,9 @@ def convert_binary_elementwise(
     # tensor. This is safe because broadcast will pad dimensions on the left
     # (prepend) to make lhs and rhs shape compatible.
     if network.has_implicit_batch_dimension:
-        if isinstance(lhs_val, torch.Tensor):
```
Comment (apbose) [inline py/torch_tensorrt/dynamo/conversion/impl/elementwise/base.py:119]: Can we remove the `network.has_implicit_batch_dimension` altogether here?

File: py/torch_tensorrt/dynamo/conversion/impl/elementwise/base.py [hunk 4]
```diff
@@ -103,9 +104,13 @@ def convert_binary_elementwise(
     # dtype but we don't have a way to detect whether it makes sense for the
     # scalar to be float or half. Hence we go with the lhs dtype.
     if is_lhs_trt_tensor and isinstance(rhs_val, (float, int)):
-        rhs_val = torch.tensor([rhs_val], dtype=lhs_dtype)
```
Comment (apbose) [inline py/torch_tensorrt/dynamo/conversion/impl/elementwise/base.py:106]: Why is this getting converted to `np` array in place of the `torch.Tensor`. Is that because the `torch.Tensor` was leading to clones getting produced in the graph?

File: py/torch_tensorrt/dynamo/conversion/converter_utils.py [hunk 5]
```diff
@@ -43,27 +43,49 @@ def get_node_name(node: torch.fx.Node) -> str:
...
+        node.op == "call_function"
+        and any(
+            arg.op == "placeholder"
+            for arg in node.args
+            if isinstance(arg, torch.fx.Node)
+        )
+        and any(user.op == "output" for user in list(node.users.keys()))
+    )
+
+
```
Comment (apbose) [inline py/torch_tensorrt/dynamo/conversion/converter_utils.py:59]: Why is there `any` used here? I believe that the above is for `to_copy` and` clone` cases having one node arg only? Or are there different cases for this?

File: py/torch_tensorrt/dynamo/conversion/impl/elementwise/base.py (hunk 4 above)
Comment (gs-olive) [reply py/torch_tensorrt/dynamo/conversion/impl/elementwise/base.py:106]: Creating `torch.Tensor` objects within the `torch.compile` scope causes them to be "Fake-ified" which removes the data contained within them.

File: py/torch_tensorrt/dynamo/conversion/converter_utils.py (hunk 5 above)
Comment (gs-olive) [reply py/torch_tensorrt/dynamo/conversion/converter_utils.py:59]: `any` is used here because we are checking if any of the users of the node are outputs, meaning that the node is the only function between a placeholder and an output. We are effectively search for subgraphs where an input is followed by a function is followed by an output.

File: py/torch_tensorrt/dynamo/conversion/impl/elementwise/base.py (hunk 3 above)
Comment (gs-olive) [reply py/torch_tensorrt/dynamo/conversion/impl/elementwise/base.py:119]: Sure, I will make this change

**Review by apbose — APPROVED** [review]
